spark_board/plan_extractor/heuristics.py

```diff
@@ -0,0 +1,94 @@
+from .transformations_dag import Metadata, TransformationColumn, TransformationNode, TransformationType, Condition
+
+from typing import Dict, List, Set, TypeVar
+
+
+class Heuristic(object):
+
+    def can_apply(self, dag: TransformationNode) -> bool:
+        raise NotImplementedError("Abstract method")
+    
+    def apply(self, original: TransformationNode) -> TransformationNode:
+        raise NotImplementedError("Abstract method")
+
+
+class MergeJoinAndProject(Heuristic):
+
+    def __init__(self) -> None:
+        self.already_merged_joins: Set[TransformationNode] = set()
+
+    def can_apply(self, dag: TransformationNode) -> bool:
+        if dag.type != TransformationType.Project:
+            return False
+        if len(dag.children) != 1:
+            return False
+        if dag.children[0].type != TransformationType.Join:
+            return False
+        if dag.children[0].metadata["join_type"] == "Cross":
```

Comment:
are you 100% sure this is the property yo check?
If I recall correctly, in pyspark the columns are "merged" when you join like:

```python
df1.join(df2, on=["col1", "col2"])
```

if you do, for example:

```python
df1.join(df2, on=[(df1.col1 == df2.col1) & (df1.col2 == df2.col2)])
```

you would have both `col1` and `col2` repeated.

Reply:
Yes, there is a `DataFrame#crossJoin` operation that doesn't have a condition, and it doesn't add the Project node, but the transformation type is still a Join. This condition avoids applying the heuristic in that incorrect scenario.
However, I will add more tests with those cases you're mentioning